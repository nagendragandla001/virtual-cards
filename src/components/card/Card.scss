.card-container {
  transition: ease-in-out;
  animation: slideMe 0.5s ease-in;
  padding: 1em;
  border-radius: 8px;
  width: 300px;
  height: 130px;
  gap: 0.5em;
  display: flex;
  justify-content: center;
  flex-direction: column;
  align-items: flex-start;
  color: #fff;
  position: relative;

  & .logo {
    position: absolute;
    bottom: 1em;
    right: 1em;
    display: flex;
    & .logo-1 {
      height: 2em;
      width: 2em;
      border-radius: 50%;
      background: #ec001b;
    }
    & .logo-2 {
      height: 2em;
      width: 2em;
      border-radius: 50%;
      background: #f79e1c;
      margin-left: -1em;
    }
  }
}

@keyframes slideMe {
  0% {
    transform: translate(-20px);
  }
  50% {
    transform: translate(-10px);
  }
  100% {
    transform: translate(0px);
  }
}
