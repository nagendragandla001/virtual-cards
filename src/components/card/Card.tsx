import { CardProps } from "../../lib/Reducers";
import { CardState } from "../../lib/Context";
import { useState } from "react";
import { AiFillEdit } from "react-icons/ai";
import "./Card.scss";


const Card = ({ card_name, card_last_four, expiry, colour }: CardProps): JSX.Element => {
  const { dispatch } = CardState();

  const [toggleState, setToggleState] = useState({
    card_name: false,
    card_last_four: false,
    expiry: false,
  });

  const handleChange = (key: string, value: string): void => {
    dispatch({
      type: "UPDATE_CARD",
      payload: { [key]: value },
    });
  };

  const updateToggleState = (key: string, value: boolean): void => {
    setToggleState((prev) => ({
      ...prev,
      [key]: value,
    }));
  };

  return (
    <section className="card-container"
      data-testid={`card-${card_name}`}
      style={{ background: colour || '#131a30' }}>
      {toggleState.card_name ? (
        <input
          type="text"
          value={card_name}
          onChange={(e) => handleChange("card_name", e.target.value)}
          onBlur={(): void => updateToggleState("card_name", false)}
          required
        />
      ) : (
        <span onDoubleClick={(): void => updateToggleState("card_name", true)}>
          {card_name ? (
            card_name
          ) : (
            <span>
              Name <AiFillEdit />
            </span>
          )}
        </span>
      )}
      {toggleState.card_last_four ? (
        <div>
          <span>**** **** ****</span>
          <input
            type="text"
            value={card_last_four}
            onChange={(e) => handleChange("card_last_four", e.target.value)}
            onBlur={(): void => updateToggleState("card_last_four", false)}
          />
        </div>
      ) : (
        <div
          onDoubleClick={(): void => updateToggleState("card_last_four", true)}
        >
          **** **** ****{" "}
          {card_last_four ? (
            card_last_four
          ) : (
            <span>
              {" "}
              <AiFillEdit className="cursor-pointer" />
            </span>
          )}
        </div>
      )}
      <div>
        <label>Expiry</label>
        {toggleState.expiry ? (
          <div>
            <input
              type="text"
              value={expiry}
              onChange={(e) => handleChange("expiry", e.target.value)}
              onBlur={(): void => updateToggleState("expiry", false)}
            />
          </div>
        ) : (
          <div onDoubleClick={(): void => updateToggleState("expiry", true)}>
            {expiry}
          </div>
        )}
      </div>
      <div className="logo">
        <div className="logo-1"></div>
        <div className="logo-2"></div>
      </div>
    </section>
  );
};

export default Card;
