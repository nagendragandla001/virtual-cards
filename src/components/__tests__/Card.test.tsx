import { render, screen, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Card from '../card/Card';

afterEach(() => {
    cleanup();
});

test('Shoudl render the Card component', () => {
    const card = { card_name: "Nagendra Gandla", card_last_four: "9889", expiry: "09/23", colour: "#ffffff" };

    render(<Card
        card_name={card.card_name}
        card_last_four={card.card_last_four}
        expiry={card.expiry}
        colour={card.colour} />);

    const cardElement = screen.getByTestId('card-Nagendra Gandla');
    expect(cardElement).toBeInTheDocument();
    expect(cardElement).toHaveTextContent('Nagendra Gandla');
    expect(cardElement).toHaveStyle({ background: '#ffffff' });
});

test('matches snapshots', () => {
    const card = { card_name: "Nagendra Gandla", card_last_four: "9889", expiry: "09/23", colour: "#ffffff" };
    const tree = renderer.create(<Card
        card_name={card.card_name}
        card_last_four={card.card_last_four}
        expiry={card.expiry}
        colour={card.colour} />).toJSON();

    expect(tree).toMatchSnapshot();
});