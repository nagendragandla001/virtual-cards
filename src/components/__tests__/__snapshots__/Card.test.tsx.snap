// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshots 1`] = `
<section
  className="card-container"
  data-testid="card-Nagendra Gandla"
  style={
    Object {
      "background": "#ffffff",
    }
  }
>
  <span
    onDoubleClick={[Function]}
  >
    Nagendra Gandla
  </span>
  <div
    onDoubleClick={[Function]}
  >
    **** **** ****
     
    9889
  </div>
  <div>
    <label>
      Expiry
    </label>
    <div
      onDoubleClick={[Function]}
    >
      09/23
    </div>
  </div>
  <div
    className="logo"
  >
    <div
      className="logo-1"
    />
    <div
      className="logo-2"
    />
  </div>
</section>
`;
